shared/src/subgraph.rs

```diff
@@ -0,0 +1,209 @@
+//! A module implementing a client for querying subgraphs.
+
+use crate::http::default_http_client;
+use anyhow::{bail, Result};
+use lazy_static::lazy_static;
+use reqwest::{Client, IntoUrl, Url};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use thiserror::Error;
+
+/// A general client for querying subgraphs.
+pub struct SubgraphClient {
+    client: Client,
+    subgraph_url: Url,
+}
+
+lazy_static! {
+    pub static ref DEFAULT_GRAPH_API_BASE_URL: Url =
+        Url::parse("https://api.thegraph.com/subgraphs/name")
+            .expect("invalid default Graph API base URL");
+}
+
+impl SubgraphClient {
+    /// Creates a new subgraph client from the specified organization and name.
+    pub fn new(org: impl AsRef<str>, name: impl AsRef<str>) -> Result<Self> {
+        Self::with_base_url(DEFAULT_GRAPH_API_BASE_URL.clone(), org, name)
+    }
+
+    /// Creates a new subgraph client with the specified base URL.
+    pub fn with_base_url(
+        base_url: impl IntoUrl,
+        org: impl AsRef<str>,
+        name: impl AsRef<str>,
+    ) -> Result<Self> {
+        let subgraph_url = base_url
+            .into_url()?
+            .join(org.as_ref())?
+            .join(name.as_ref())?;
+        Ok(Self {
+            client: default_http_client()?,
+            subgraph_url,
+        })
+    }
+
+    /// Performs the specified GraphQL query on the current subgraph.
+    pub async fn query<T>(&self, q: impl AsRef<str>) -> Result<T>
+    where
+        T: DeserializeOwned,
+    {
+        self.client
+            .post(self.subgraph_url.clone())
+            .json(&Query { query: q.as_ref() })
+            .send()
+            .await?
+            .json::<QueryResponse<T>>()
+            .await?
+            .into_result()
+    }
+}
+
+/// A GraphQL query.
+#[derive(Serialize)]
+struct Query<'a> {
+    query: &'a str,
+}
```

Comment:
Is this something that will have more content later or will we always allow queries to be arbitrary strings? This leads me to believe that graph queries have no common structure...

Reply:
I just don't want to write a GraphQL parser.